internal/configs/virtualserver.go

```diff
@@ -372,11 +398,18 @@ func (vsc *virtualServerConfigurator) GenerateVirtualServerConfig(vsEx *VirtualS
 	matchesRoutes := 0
 
 	variableNamer := newVariableNamer(vsEx.VirtualServer)
+	upstreamNamer := newUpstreamNamerForVirtualServer(vsEx.VirtualServer)
 
 	// generates config for VirtualServer routes
 	for _, r := range vsEx.VirtualServer.Spec.Routes {
+		errorPages := r.ErrorPages
+		upstreamName := upstreamNamer.GetNameForUpstreamFromAction(r.Action)
+		if upstreamName != "" && r.ErrorPages != nil {
+			upstream := crUpstreams[upstreamName]
+			errorPages = checkGrpcErrorPageCodes(r.ErrorPages, isGRPC(upstream.Type), upstreamName, vsc, vsEx)
+		}
 		errorPageIndex := len(errorPageLocations)
-		errorPageLocations = append(errorPageLocations, generateErrorPageLocations(errorPageIndex, r.ErrorPages)...)
+		errorPageLocations = append(errorPageLocations, generateErrorPageLocations(errorPageIndex, errorPages)...)
```

Comment:
I wonder what you think about the following alternative. Instead having additional logic to check the error pages and upstreams here, move it down to the `generateLocation`. In that case, `generateLocation` and a few more functions will have to be changed. However, using `generateLocation` will make sure the cases when upstreams are in `splits`, `matches` are covered: `generateLocation` has the complete information to handle all the cases.


For example, `generateLocation` can be extended as follows:
```
type errorPageDetails struct {
	pages []conf_v1.ErrorPage # the current errorPages parameter of the generateLocation
	index int # errPageIndex
	owner runtime.Object # this will be either VS or VSR - the resource that owns the error pages.
}

func generateLocation( ..., errorPageDetails *errorPageDetails, ...) (..., Warnings){
	// if upstream is gRPC and error pages include non-compatible code
	// generate an error
       
       . . .
}
```

note that in this approach, the function will also return Warnings, which can be processed in the `GenerateVirtualServerConfig`.

```
loc, returnLoc, warnings := generateLocation(...)
vsc.warnings.Add(warnings)
```